crates/accelerate/src/py_sparse_observable.rs

```diff
@@ -0,0 +1,2409 @@
+/// * negate all the coefficients in an observable with unary ``-``
+///
+/// Each of the basic binary arithmetic operators has a corresponding specialized in-place method,
+/// which mutates the left-hand side in-place.  Using these is typically more efficient than the
+/// infix operators, especially for building an observable in a loop.
+///
+/// The tensor product is calculated with :meth:`tensor` (for standard, juxtaposition ordering of
+/// Pauli labels) or :meth:`expand` (for the reverse order).  The ``^`` operator is overloaded to be
+/// equivalent to :meth:`tensor`.
+///
+/// .. note::
+///
+///     When using the binary operators ``^`` (:meth:`tensor`) and ``&`` (:meth:`compose`), beware
+///     that `Python's operator-precedence rules
+///     <https://docs.python.org/3/reference/expressions.html#operator-precedence>`__ may cause the
+///     evaluation order to be different to your expectation.  In particular, the operator ``+``
+///     binds more tightly than ``^`` or ``&``, just like ``*`` binds more tightly than ``+``.
+///
+///     When using the operators in mixed expressions, it is safest to use parentheses to group the
+///     operands of tensor products.
+///
+/// A :class:`SparseObservable` has a well-defined :meth:`adjoint`.  The notions of scalar complex
+/// conjugation (:meth:`conjugate`) and real-value transposition (:meth:`transpose`) are defined
+/// analogously to the matrix representation of other Pauli operators in Qiskit.
+///
+///
+/// Efficiency notes
+/// ----------------
+///
+/// Internally, :class:`SparseObservable` is in-place mutable, including using over-allocating
+/// growable vectors for extending the number of terms.  This means that the cost of appending to an
+/// observable using ``+=`` is amortised linear in the total number of terms added, rather than
+/// the quadratic complexity that the binary ``+`` would require.
+///
+/// Additions and subtractions are implemented by a term-stacking operation; there is no automatic
+/// "simplification" (summing of like terms), because the majority of additions to build up an
+/// observable generate only a small number of duplications, and like-term detection has additional
+/// costs.  If this does not fit your use cases, you can either periodically call :meth:`simplify`,
+/// or discuss further APIs with us for better building of observables.
+#[pyclass(name = "SparseObservable", module = "qiskit.quantum_info", sequence)]
+#[derive(Debug)]
+pub struct PySparseObservable {
+    // This class keeps a pointer to a pure Rust-SparseTerm and serves as interface from Python.
+    inner: Arc<RwLock<SparseObservable>>,
+}
+
+impl Clone for PySparseObservable {
+    // we need to implement clone specifically to ensure the underlying inner data
+    // is cloned, not the reference to it
+    fn clone(&self) -> Self {
+        let inner = self.inner.read().unwrap();
+        Self {
+            inner: Arc::new(RwLock::new(inner.clone())),
+        }
+    }
+}
```

Comment:
The reason is that we didn't think of putting this into `__copy__` and we wanted to ensure we clone the object, not the ref 🙂 You're right, it's not really intended to be called from Rust space

> (though I think they probably should already work).

I.e. that Python's native `__copy__` would already copy the `inner` data instead of the reference to it?

Reply:
`copy.copy` and `copy.deepcopy` don't _need_ `__copy__` etc to be defined - they'll use the pickle-protocol methods (`__reduce__` or  `__getstate__`/`__statestate__`) if they're not, and that causes `SparseObservable` to integrate with them already. I'm relatively sure I wrote a test.